### PR TITLE
doc: adding more clarifications to optimizations, code

### PR DESCRIPTION
**Pull request by kupermind**
- More clarifications to optimizations, potentially incorrectly interpreted code;
- Added comments to frequently encountered false positives.

### CODE REVIEW

**Review by 77ph — APPROVED** [review]

**Review by mariapiamo — APPROVED** [review]

File: contracts/veOLAS.sol [hunk 1]
```diff
@@ -634,6 +635,8 @@ contract veOLAS is IErrors, IVotes, IERC20, IERC165 {
     }
 
     /// @dev Gets the block time adjustment for two neighboring points.
+    /// @notice `blockNumber` must not be lower than the contract deployment block number (15050278),
```
Comment (DavidMinarsch) [inline contracts/veOLAS.sol:638]:
```suggestion
    /// @notice `blockNumber` must not be lower than the contract deployment block number (15050278),
```
 (15050278) cannot be known before deploying, so remove it here.

File: docs/optimizations.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+- Custom `reverts` and no `require` statements;
+- Storage vs memory keyword: decision is on what to use after testing for gas spending;
+- Explicit type casting even where the compiler does it by default;
+- No usage of `safeTransfer()` from the OpenZeppelin (or their analogues) where it is not absolutely necessary;
+- Follow `Checks-Effects-Interactions` pattern;
+- Events are not subject to reentrancy attacks since events do not cause a function to throw;
+- Avoid reverts where the logic allows not to revert and return a zero value instead;
+- All `unchecked` statements are commented on;
+- All functions with `owner` privileges are not considered to be called from malicious actors and are not meant to be executed for other than common sense scenarios;
+- Our `OLAS` token is considered to be called `with good intentions`, and strictly corresponds to the ERC20 standard;
```
Comment (DavidMinarsch) [inline docs/optimizations.md:16]: What does "with good intentions" mean? Not sure what you're trying to say here.

File: docs/optimizations.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+- Utilize standard code-generated reverts and do not duplicate them, especially in places where the contract itself or any external contracts lead to any negative or false scenario;
+- Reverts are given the security priority and the need of a business logic, such that they can be omitted where possible to save on gas;
```
Comment (DavidMinarsch) [inline docs/optimizations.md:18]: This sentence is not very clear

File: docs/optimizations.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+- Pure storage variables are normally memory-cached to undergo operations, and are assigned back to the storage slot afterwards;
+- Prioritize lower consuming gas operations when considering statements (i.e., `(i + 1) > j` instead of `i >= j`);
+- Use addresses instead of explicit interface variables;
+- Minimize the number of inherited contracts;
+- Minimize the nesting level of function calls;
+- `enum`-s are used for a better code readability;
+- Find tradeoff between optimizations and a code readability;
+- When implementing standards (i.e., `ERC20`), the most optimal well-known solutions are utilized;
+- When implementing non-standards, the most gas optimized and logical solutions are chosen;
+- Function returns sometimes mean the correct workflow of the function, and do not bear a promise that variables inside are modified.
```
Comment (DavidMinarsch) [inline docs/optimizations.md:28]: Not sure I understand this sentence fully

File: docs/optimizations.md (hunk 2 above)
Comment (kupermind) [reply docs/optimizations.md:16]: It means that the token code does not do contain any special harmful code. Will rephrase

File: docs/optimizations.md (hunk 3 above)
Comment (kupermind) [reply docs/optimizations.md:18]: Rephrased

File: docs/optimizations.md (hunk 4 above)
Comment (kupermind) [reply docs/optimizations.md:28]: Rephrased

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**